ciao-controller/command.go

```diff
@@ -266,65 +306,29 @@ func (c *controller) startWorkload(w types.WorkloadRequest) ([]*types.Instance,
 		if w.Subnet == "" {
 			newIP = IPPool[i]
 		}
+
+		name := w.Name
+		if name != "" {
+			if w.Instances > 1 {
+				name = fmt.Sprintf("%s-%d", name, i)
+			}
+		}
+
 		wg.Add(1)
-		go func(newIP net.IP) {
+		go func(newIP net.IP, name string) {
 			sem <- 1
 			defer func() {
 				<-sem
 				wg.Done()
 			}()
 
-			startTime := time.Now()
-
-			name := w.Name
-			if name != "" {
-				if w.Instances > 1 {
-					name = fmt.Sprintf("%s-%d", name, i)
-				}
-			}
-
-			instance, err := newInstance(c, w.TenantID, &wl, w.Volumes, name, w.Subnet, newIP)
+			instance, err := c.createInstance(w, wl, name, newIP)
 			if err != nil {
 				e = errors.Wrap(err, "Error creating instance")
```

Comment:
I think this assignment is racy as e is declared outside the closure.  If multiple instances fail at the same time you'll have a race

Reply:
You could replace the waitgroup with a chan error.  Each go routine writes to the channel when there's an error or when they're finished.  Instead of wg.Wait you have

```
for i := 0; i < w.Instances; i++ {
   err := <-errChan
  if e == nil {
    e = err
  }
}
```

This returns the first error.

Reply:
Also the first loop is wrong

 	for i := 0; i < w.Instances && e == nil; i++ {


There's no point checking e == nil here.  It's racy